Document .NET Core resource fallback process

sdmaclea opened the pull request:
@rpetrusha Ultimately this belongs in the .NET Guide as you mentioned.  In the interim adding .NET Core doc here...

This is a rough first attempt.  Feel free to recommend edits...  I chose some content replication rather that restructuring to make the content digestible.

## Summary

Document .NET Core resource fallback process

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/framework/resources/packaging-and-deploying-resources-in-desktop-apps?branch=pr-en-us-8920)

sdmaclea committed:
Document .NET Core resource fallback process

@rpetrusha Ultimately this belongs in the .NET Guide as you mentioned.  In the interim adding .NET Core doc here...

This is a rough first attempt.  Feel free to recommend edits...  I chose some content replication rather that restructuring to make the content digestible.

rpetrusha commented:
@tarekgh and @krwq can you review this?

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -177,7 +177,29 @@ vbc Example1.vb
+  
+3. The runtime next searches the parent culture assemblies through many potential levels, each time repeating steps 1 & 2. 
+
+The parent culture is defined as the appropriate fallback culture. Consider parents as fallback candidates, because providing any resource is preferable to throwing an exception. This process also allows you to reuse resources. You should include a particular resource at the parent level only if the child culture doesn't need to localize the requested resource. For example, if you supply satellit ...
```
- This needs to be indented two spaces to continue to numbered list item. 
- the appropriate fallback culture --> an appropriate fallback culture.

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -177,7 +177,29 @@ vbc Example1.vb
+4. If the culture that was originally specified and all parents have been searched and the resource is still not found, the resource for the default (fallback) culture is used. Typically, the resources for the default culture are included in the main application assembly. However, you can specify a value of <xref:System.Resources.UltimateResourceFallbackLocation.Satellite> for the <xref:System.Re ...
+  
+    > [!NOTE]
+    >  The default resource is the only resource that can be compiled with the main assembly. Unless you specify a satellite assembly by using the <xref:System.Resources.NeutralResourcesLanguageAttribute> attribute, it is the ultimate fallback (final parent). Therefore, we recommend that you always include a default set of resources in your main assembly. This helps prevent exceptions from being  ...
```
A comma should follow "By including a default resource file"
no comma before "and ensure"

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -177,7 +177,29 @@ vbc Example1.vb
+  
+5. Finally, if the runtime doesn't find a resource for a default (fallback) culture, a <xref:System.Resources.MissingManifestResourceException> or <xref:System.Resources.MissingSatelliteAssemblyException> exception is thrown to indicate that the resource could not be found.  
+  
+ For example, suppose the application requests a resource localized for Spanish (Mexico) (the es-MX culture). The runtime first searches the global assembly cache for the assembly that matches es-MX, but doesn't find it. The runtime then searches the directory of the currently executing assembly for an es-MX directory. Failing that, the runtime searches the global assembly cache again for a paren ...
```
Does this (especially the global assembly cache) apply to .NET Core?

jkotas replied:
It does not. GAC (Global assembly cache) does not exist in .NET Core.

rpetrusha replied:
Ah, this is just a copy-and-paste error, @sdmaclea -- the paragraph is repeated earlier on line 95. So this paragraph should be removed. Given that .NET Core's probing is simpler than .NET Framework's, I don't think that we need a for example paragraph -- the steps in the probe are sufficient.

rpetrusha reviewed (changes requested):
Thanks for adding this section on resource fallback for .NET Core, @sdmaclea. I've left some mostly minor comments, along with some more general comments on changes to restructure the topic slightly, which are necessitated by the addition of the .NET Core section. I've also asked @tarekgh and @krwq  to review it.

tarekgh commented:
Thanks @sdmaclea for the writing. it looks good to me. 

Could you please add a note regarding when searching for satellite assemblies on Linux/OSX, the search is case sensitive. i.e. the language folder name has to be in same casing as the culture name (i.e. es-MX and not something like es-mx). I hope we can fix this issue in the frameowrk to make this search case insensitive.

tarekgh commented:
other than that, LGTM.

tarekgh reviewed (approved):
left small comment I hope you can address it before your merge

vitek-karas commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -177,7 +177,29 @@ vbc Example1.vb
   
+1.  The runtime checks the directory of the currently executing assembly for a directory that matches the requested culture. If it finds the directory, it searches that directory for a valid satellite assembly for the requested culture. The runtime then searches the satellite assembly for the requested resource. If it finds the resource in the assembly, it uses it. If it doesn't find the resource ...
+  
+2.  The runtime raises the <xref:System.Runtime.Loader.AssemblyLoadContext.Resolving?displayProperty=nameWithType> event to indicate that it is unable to find the satellite assembly. If you choose to handle the event, your event handler can return a reference to the satellite assembly whose resources will be used for the lookup. Otherwise, the event handler returns `null` and the search continues ...
```
I haven't checked the code, but I'm pretty sure the runtime issues a normal "bind-by-name" - basically `Assembly.Load` equivalent. Which would trigger entire series of things (`AssemblyLoadContext.Load`, then fallback to default, then the default's and custom ALC's `Resolving` event and finally the `AppDomain` resolve event).
This also includes the existing search in the resource probing paths as specified by the host (based on `.deps.json`).

sdmaclea replied:
Most of what you say is true.  Based on one of our earlier discussions, I had try to avoid unnecessary complexity here.  I over simplified.  Partially because I was focused on pruning the .NET Framework algorithm.

The code is calling into AssemblyNative::Load().  The arguments for satellite load are different than a normal assembly load so the same full normal algorithm is not called.  I will step through the native code to check some of your assertions to make sure this description is as accurate and simple to understand as possible.

sdmaclea committed:
PR feedback

sdmaclea commented:
@rpetrusha @vitek-karas  I believe I have addressed all PR feedback.  I have reviewed the preview document link in the OP and am happy with the current state.  Please review

vitek-karas commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,39 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+### .NET Core resource fallback process
+ The .NET Core resource fallback process involves the following steps:
+
+1.  The runtime attempts to load a satellite assembly for the requested culture, eventually calling <xref:System.Runtime.Loader.AssemblyLoadContext.Load(System.Reflection.AssemblyName)?displayProperty=nameWithType> on the AssemblyLoadContext containing the the currently executing assembly.  The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType>:
```
> the the currently executing assembly

Remove the duplicate "the"

vitek-karas replied:
Should the AssemblyLoadContext be marked as code? - that is `AssemblyLoadContext`.

vitek-karas replied:
Also I don't really follow the logic of this paragraph: it is basically
> The runtime eventually calls ALC.Load on the right context. The ALC.Default does this:

There doesn't seem to be any logic connection between ALC.Load and the immediately following description of ALC.Default.

I think it should say something like: If the AssemblyLoadContext is the default one or it falls back to the default, then the runtime will:

vitek-karas reviewed (approved):
Other than the small requested tweaks - LGTM.

sdmaclea committed:
Further PR feedback

sdmaclea commented:
@rpetrusha ping

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+     * Checks the directory of the currently executing assembly for a subdirectory that matches the requested culture. If it finds the subdirectory, it searches that subdirectory for a valid satellite assembly for the requested culture and loads it.
+
+       > [!NOTE]
+       >  On operating systems with case sensistive file systems (e.g. Linux and OSX) the culture name subdirectory search is case sensitive.  The subdirectory name must exactly match the case of the <xref:System.Globalization.CultureInfo.Name?displayProperty=nameWithType> i.e. `es` or `es-MX`.
```
- case sensistive --> case-sensitive
- (e.g., Linux and OSX) --> (that is, Linux and macOS), (note the comma)
- i.e \`es\` or \`es-MX\` --> (for example, \`es\` or \`es-MX\`)

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+       >  On operating systems with case sensistive file systems (e.g. Linux and OSX) the culture name subdirectory search is case sensitive.  The subdirectory name must exactly match the case of the <xref:System.Globalization.CultureInfo.Name?displayProperty=nameWithType> i.e. `es` or `es-MX`.
+
+       > [!NOTE]
+       > If the programmer has derived a custom assembly load context from <xref:System.Runtime.Loader.AssemblyLoadContext> the situation is complicated.  If the executing assembly was loaded into the custom context the runtime will load the satellite into the custom context.  The details are out of scope for this document.  See  <xref:System.Runtime.Loader.AssemblyLoadContext>.
```
comma after:
- the xref to AssemblyLoadContext
- "into the custom context
will load --> loads

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+     * If a satellite has not been found, the <xref:System.Runtime.Loader.AssemblyLoadContext> raises the <xref:System.Runtime.Loader.AssemblyLoadContext.Resolving?displayProperty=nameWithType> event to indicate that it is unable to find the satellite assembly. If you choose to handle the event, your event handler can load and return a reference to the satellite assembly.
+     * If a satellite still has not been found, the AssemblyLoadContext causes the AppDomain to trigger an <xref:System.AppDomain.AssemblyResolve?displayProperty=nameWithType> event to indicate that it is unable to find the satellite assembly. If you choose to handle the event, your event handler can load and return a reference to the satellite assembly.
+
+2. If a Satellite is found, the runtime then searches the satellite assembly for the requested resource. If it finds the resource in the assembly, it uses it. If it doesn't find the resource, it continues the search.
```
Satellite --> satellite assembly
searches the satellite assembly --> searches it

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+       > [!NOTE]
+       > If the programmer has derived a custom assembly load context from <xref:System.Runtime.Loader.AssemblyLoadContext> the situation is complicated.  If the executing assembly was loaded into the custom context the runtime will load the satellite into the custom context.  The details are out of scope for this document.  See  <xref:System.Runtime.Loader.AssemblyLoadContext>.
+
+     * If a satellite has not been found, the <xref:System.Runtime.Loader.AssemblyLoadContext> raises the <xref:System.Runtime.Loader.AssemblyLoadContext.Resolving?displayProperty=nameWithType> event to indicate that it is unable to find the satellite assembly. If you choose to handle the event, your event handler can load and return a reference to the satellite assembly.
```
satellite --> satellite assembly (here and later)

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+
+3. The runtime next searches the parent culture assemblies through many potential levels, each time repeating steps 1 & 2.
+
+     The parent culture is defined as an appropriate fallback culture. Consider parents as fallback candidates, because providing any resource is preferable to throwing an exception. This process also allows you to reuse resources. You should include a particular resource at the parent level only if the child culture doesn't need to localize the requested resource. For example, if you supply sate ...
```
would contain --> contains
could provide --> provides

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+
+     The parent culture is defined as an appropriate fallback culture. Consider parents as fallback candidates, because providing any resource is preferable to throwing an exception. This process also allows you to reuse resources. You should include a particular resource at the parent level only if the child culture doesn't need to localize the requested resource. For example, if you supply sate ...
+
+     Each culture has only one parent, which is defined by the <xref:System.Globalization.CultureInfo.Parent%2A?displayProperty=nameWithType> property, but a parent might have its own parent. The search for parent cultures stops when a culture's <xref:System.Globalization.CultureInfo.Parent%2A> property returns <xref:System.Globalization.CultureInfo.InvariantCulture%2A?displayProperty=nameWithTyp ...
```
no semicolon before "for resource fallback". This should be a new sentence.

sdmaclea replied:
Done

rpetrusha commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -122,10 +123,44 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
+
+     Each culture has only one parent, which is defined by the <xref:System.Globalization.CultureInfo.Parent%2A?displayProperty=nameWithType> property, but a parent might have its own parent. The search for parent cultures stops when a culture's <xref:System.Globalization.CultureInfo.Parent%2A> property returns <xref:System.Globalization.CultureInfo.InvariantCulture%2A?displayProperty=nameWithTyp ...
+
+4. If the culture that was originally specified and all parents have been searched and the resource is still not found, the resource for the default (fallback) culture is used. Typically, the resources for the default culture are included in the main application assembly. However, you can specify a value of <xref:System.Resources.UltimateResourceFallbackLocation.Satellite> for the <xref:System.Re ...
```
- UltimateResourceFallbackLocation.Satellite should be ?displayProperty.nameWithType
- NeutralResourcesLanguageAttribute.Location property of the ...attribute -- can be shortened to \<xref:System.Resources.NeutralResourcesLanguageAttribute.Location?displayProperty=nameWithType> property
- no comma before "rather"

sdmaclea replied:
Done

rpetrusha reviewed (changes requested):
I apologize for my belated review, @sdmaclea. This looks really good. I've suggested a number of minor changes.

sdmaclea committed:
More PR feedback

sdmaclea commented on docs/framework/resources/packaging-and-deploying-resources-in-desktop-apps.md:
```diff
@@ -58,24 +58,25 @@ Applications rely on the .NET Framework Resource Manager, represented by the <xr
  To improve lookup performance, apply the <xref:System.Resources.NeutralResourcesLanguageAttribute> attribute to your main assembly, and pass it the name of the neutral language that will work with your main assembly.  
   
+### .NET Framework resource fallback process
+ The .NET Framework resource fallback process is involves the following steps:
```
done

sdmaclea commented:
Thanks @rpetrusha  I believe I have made all your requested changes.
